### PR TITLE
#4 Implement pgvector backend

### PR DESCRIPTION
**Pull request by tm-kn**
#4 

An initial stab at the pgvector backend support.

### CODE REVIEW

File: src/wagtail_vector_index/backends/pgvector/migrations/0001_enable_pgvector_extension.py [hunk 1]
```diff
@@ -0,0 +1,8 @@
+# Generated by Django 4.2.7 on 2023-11-22 14:46
+
+from django.db import migrations
+from pgvector.django import VectorExtension
+
+
+class Migration(migrations.Migration):
+    operations = [VectorExtension()]
```
Comment (tm-kn) [inline src/wagtail_vector_index/backends/pgvector/migrations/0001_enable_pgvector_extension.py:8]: This bizzarely does not fail on SQLite... ;(

File: src/wagtail_vector_index/backends/pgvector/migrations/0002_initial.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+                (
+                    "id",
+                    models.AutoField(
+                        auto_created=True,
+                        primary_key=True,
+                        serialize=False,
+                        verbose_name="ID",
+                    ),
+                ),
+                ("vector", pgvector.django.VectorField()),
```
Comment (tm-kn) [inline src/wagtail_vector_index/backends/pgvector/migrations/0002_initial.py:29]: Does not fail on SQLite either.

**Review by tomusher — COMMENTED** [review]
This looks really great to me. I think we can get this in as-is and refine if needed once we start testing internally.

File: src/wagtail_vector_index/backends/pgvector/backend.py [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+    ...
+
+
+class PgvectorIndex(Index):
+    class DistanceMethod(enum.Enum):
+        EUCLIDEAN = "euclidean"
+        COSINE = "cosine"
+        MAX_INNER_PRODUCT = "max_inner_product"
+
+    upsert_batch_size: int = 500
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply src/wagtail_vector_index/backends/pgvector/backend.py:25]: Also abstraction for the per-index settings don't exist yet, so it's likely impossible.

File: src/wagtail_vector_index/backends/pgvector/backend.py [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+    def clear(self):
+        self._get_queryset().delete()
+
+    def similarity_search(
+        self, query_vector, *, limit: int = 5
+    ) -> Generator[SearchResponseDocument, None, None]:
+        for pgvector_embedding in (
+            self._get_queryset()
+            .select_related("embedding")
+            .annotate(distance=self._distance_method_cls()("vector", query_vector))
```
[the remark this answers is not in the document]
Comment (tm-kn) [reply src/wagtail_vector_index/backends/pgvector/backend.py]: Done. Not sure if it's any good.

**Comment by tm-kn** [comment]
> This looks really great to me. I think we can get this in as-is and refine if needed once we start testing internally.

@tomusher Do we want to get this in as-is now?

**Pull request merged**